### PR TITLE
Track document versions in the server

### LINKED ISSUE #5452: LSP: Should we be sending versions in VersionedTextDocumentIdentifier and PublishDiagnostics?
This would impact `TextDocumentEdit` (and thus `WorkspaceEdit` and our custom snippet edit) and `PublishDiagnostics`. I think the approach would be to stash the version off in a side-table at the end of the `DidChangeTextDocument` notification and grab it when publishing diagnostics or computing edits (possibly in `to_proto`).

### PR DESCRIPTION
**Pull request by kjeremy**
This also pushes diagnostics for the correct file version on close so that when it is reopened stale diagnostics are not shown.

Closes #5452

### CODE REVIEW

File: crates/rust-analyzer/src/main_loop.rs [hunk 1]
```diff
@@ -412,29 +420,38 @@ impl GlobalState {
             })?
             .on::<lsp_types::notification::DidChangeTextDocument>(|this, params| {
                 if let Ok(path) = from_proto::vfs_path(&params.text_document.uri) {
-                    assert!(this.mem_docs.contains(&path));
+                    assert!(this.mem_docs.contains_key(&path));
```
Comment (matklad) [inline crates/rust-analyzer/src/main_loop.rs:423]:
```suggestion
                    *this.mem_docs.get_mut(&path).unwrap() = params.text_document.version;
```

**Comment by lnicola** [comment]
Can we use this to avoid https://github.com/rust-analyzer/rust-analyzer/issues/4266, which happens because we try to apply `cargo check` diagnostics to a modified file?

**Comment by kjeremy** [comment]
@Inicola I hope so but this change on it's own isn't enough. I was able to reproduce the crash with this branch.

**Pull request merged**